### PR TITLE
Feature: Coin98VaultProxy for free gas user on Viction

### PR DESCRIPTION
**Pull request by c98tristan**

### CODE REVIEW

File: contracts/Coin98VaultV2Proxy.sol [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    constructor() VRC25("Coin98 Vault Proxy", "C98VP", 18) {
+    }
+    function redeem(address payable vaultAddress, uint256 eventId_, uint256 index_, uint256 timestamp_, address recipient_, uint256 receivingAmount_, uint256 sendingAmount_, bytes32[] calldata proofs) public payable {
+        EventData memory eventData = _eventDatas[eventId_];
+
+        if (sendingAmount_ > 0) {
+            IERC20(eventData.sendingToken).safeTransferFrom(msg.sender, address(this), sendingAmount_);
+            IERC20(eventData.sendingToken).safeApprove(vaultAddress, sendingAmount_);
+            IERC20(eventData.sendingToken).safeTransfer(vaultAddress, sendingAmount_);
```
Comment (imterryyy) [inline contracts/Coin98VaultV2Proxy.sol:27]: Why you send token to vault ?

File: contracts/Coin98VaultV2Proxy.sol [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import "./lib/VRC25.sol";
+import "./Coin98VaultV2.sol";
+import "./lib/Token.sol";
+
+contract Coin98V2VaultProxy is VRC25 {
+    using AdvancedERC20 for IERC20;
+
+    mapping(uint256 => EventData) private _eventDatas;
+
+    struct EventData {
```
Comment (imterryyy) [inline contracts/Coin98VaultV2Proxy.sol:12]: Please format code

File: contracts/Coin98VaultV2Proxy.sol [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    uint8 isActive;
+  }
+
+    constructor() VRC25("Coin98 Vault Proxy", "C98VP", 18) {
+    }
+    function redeem(address payable vaultAddress, uint256 eventId_, uint256 index_, uint256 timestamp_, address recipient_, uint256 receivingAmount_, uint256 sendingAmount_, bytes32[] calldata proofs) public payable {
+        EventData memory eventData = _eventDatas[eventId_];
+
+        if (sendingAmount_ > 0) {
+            IERC20(eventData.sendingToken).safeTransferFrom(msg.sender, address(this), sendingAmount_);
```
Comment (imterryyy) [inline contracts/Coin98VaultV2Proxy.sol:25]: please check `sendingToken != address(0)`

File: contracts/Coin98VaultV2Proxy.sol [hunk 4]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+import "./lib/VRC25.sol";
+import "./Coin98VaultV2.sol";
+import "./lib/Token.sol";
+
+contract Coin98V2VaultProxy is VRC25 {
+    using AdvancedERC20 for IERC20;
+
+    mapping(uint256 => EventData) private _eventDatas;
```
Comment (imterryyy) [inline contracts/Coin98VaultV2Proxy.sol:10]: What is this ?

**Review by imterryyy — CHANGES_REQUESTED** [review]

**Review by terryyyz-coin98 — CHANGES_REQUESTED** [review]

File: tests/vault_sponsor_gas.spec.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        currentTimestamp,
+        account1Address,
+        1000000,
+        1000,
+        proofs
+      );
+    const balanceAfter = await c98Token.balanceOf(account1Address);
+    const balanceVaultAfter = await usdtToken.balanceOf(sut.address);
+
+    expect(balanceVaultAfter.sub(balanceVaultBefore).toNumber()).eq(1000);
```
Comment (terryyyz-coin98) [inline tests/vault_sponsor_gas.spec.ts:211]: use `changeTokensBalance` api to verify token transfer action.

**Review by terryyyz-coin98 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**